okhttp/src/main/java/okhttp3/internal/framed/Hpack.java

```diff
@@ -367,27 +368,125 @@ ByteString readByteString() throws IOException {
   }
 
   static final class Writer {
+    private static final byte SEPARATED_TOKEN = ':';
+    private static final int SETTINGS_HEADER_TABLE_SIZE = 4096;
+
     private final Buffer out;
+    private final Map<ByteString, Integer> headerStringToDynamicIndex =
+        new HashMap<ByteString, Integer>();
+
+    private int headerTableSizeSetting;
+    private int maxDynamicTableByteCount;
+    // Visible for testing.
+    Header[] dynamicTable = new Header[8];
+    // Array is populated back to front, so new entries always have lowest index.
+    int nextHeaderIndex = dynamicTable.length - 1;
+    int headerCount = 0;
+    int dynamicTableByteCount = 0;
 
     Writer(Buffer out) {
+      this(SETTINGS_HEADER_TABLE_SIZE, out);
+    }
+
+    Writer(int headerTableSizeSetting, Buffer out) {
+      this.headerTableSizeSetting = headerTableSizeSetting;
+      this.maxDynamicTableByteCount = headerTableSizeSetting;
       this.out = out;
     }
 
+    ByteString getHeaderString(Header entry) {
+      byte[] ret = new byte[entry.name.size() + 1 + entry.value.size()];
+      System.arraycopy(entry.name.toByteArray(), 0, ret, 0, entry.name.size());
+      ret[entry.name.size()] = SEPARATED_TOKEN;
+      System.arraycopy(entry.value.toByteArray(), 0, ret, entry.name.size() + 1,
+          entry.value.size());
+      return ByteString.of(ret);
+    }
+
+    private void clearDynamicTable() {
+      Arrays.fill(dynamicTable, null);
+      headerStringToDynamicIndex.clear();
+      nextHeaderIndex = dynamicTable.length - 1;
+      headerCount = 0;
+      dynamicTableByteCount = 0;
+    }
+
+    /** Returns the count of entries evicted. */
+    private int evictToRecoverBytes(int bytesToRecover) {
+      int entriesToEvict = 0;
+      if (bytesToRecover > 0) {
+        // determine how many headers need to be evicted.
+        for (int j = dynamicTable.length - 1; j >= nextHeaderIndex && bytesToRecover > 0; j--) {
+          bytesToRecover -= dynamicTable[j].hpackSize;
+          dynamicTableByteCount -= dynamicTable[j].hpackSize;
+          headerCount--;
+          entriesToEvict++;
+        }
+        System.arraycopy(dynamicTable, nextHeaderIndex + 1, dynamicTable,
+            nextHeaderIndex + 1 + entriesToEvict, headerCount);
+        for (Map.Entry<ByteString, Integer> p : headerStringToDynamicIndex.entrySet()) {
```

Comment:
This is simple but I wonder if there’s a datastructure better suited to this. I’m thinking about Android and the amount of boxing and unboxing that happens here.

(I guess we shouldn’t worry about micro-optimizations until we know the code is working!)